# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [55]:
import pandas as pd

In [ ]:
try:
    data = pd.read_csv('./datasets/data.csv')
except:
    data = pd.read_csv('https://***/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [56]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [57]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [58]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [59]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [60]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [61]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [62]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [63]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [64]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [65]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [66]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [67]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [68]:
data.duplicated().sum()

54

In [69]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

 <a id='header-2-6'></a>

In [70]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [71]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [72]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [73]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

<a id='task18'></a>

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [74]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [75]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Сгруппируем данные по количеству детей в семье и рассчитаем суммарное количество задолженностей по кредиту в каждой категории (по количеству детей):

In [76]:
sum_debt_ch = data.groupby('children')['debt'].sum()
display(sum_debt_ch)

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

Видно, что чем меньше детей, тем больше просрочек по выплатам кредита. Однако, для оценки зависимости между количеством детей и возвратом кредита этого недостаточно. Попробуем качественно оценить данные по каждой категории. Для этого выведем сводную таблицу с данными о суммарном количестве задолженностей (sum debt), общеем количестве взятых кредитов (count debt) и среднем значении по задолженности (mean debt), которое если умножить на 100, то получится процент задолженности в каждой категории (по количеству детей):

In [77]:
display(pd.pivot_table(data, index='children', values='debt', aggfunc=['sum', 'count', 'mean']))

sum  count      mean
          debt   debt      debt
children                       
0         1063  14107  0.075353
1          444   4809  0.092327
2          194   2052  0.094542
3           27    330  0.081818
4            4     41  0.097561
5            0      9  0.000000

Согласно данным, чем больше детей, тем выше процент просрочек, однако, категория с тремя детьми немного выпадает из этой тенденции (8,18%), а категорию с пятью детьми вообще трудно оценивать из-за малочисленности данной группы (малого количества выданных кредитов). Что интересно, несмотря на огромное количество просрочек в категории "без детей" (1063), процент задолженности минимальный (7,54%). Самый большой процент задолженности в категории "четверо детей" (9,76%).

**Вывод:** чем больше детей, тем выше риски просроченных платежей, хотя разница в процентах не столь велика, кроме того, чем меньше детей, тем люди охотнее берут кредиты.

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Сразу выведем сводную таблицу, сгрупированную по семейному положению с суммарными значениями по просроченным выплатам (sum debt), общему количеству взятых кредитов (count debt) и среднем значение по задолженности (mean debt), которое если умножить на 100, то получится процент задолженности в каждой категории (семейное положение):

In [78]:
display(pd.pivot_table(data, index='family_status', values='debt', aggfunc=['sum', 'count', 'mean']))

sum  count      mean
                      debt   debt      debt
family_status                              
Не женат / не замужем  273   2796  0.097639
в разводе               84   1189  0.070648
вдовец / вдова          63    951  0.066246
гражданский брак       385   4146  0.092861
женат / замужем        927  12266  0.075575

Самые ответственные граждане из категории "вдовец/вдова", у них минимальный процент задержки платежей (6,62%), самая безответственная категория это "не женатые/не замужние" (9,76%), что интересно категория "в разводе" имеет очень небольшой процент просрочки (7,06%).

**Вывод:** очень интересная категория, которую обязательно надо учитывать при выдаче кредита, особенно обратить внимание на категорию "не женат/не замужем".

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Сформируем сводную таблицу, сгрупированную по категориям уровня дохода, которые мы создали и распределили по ним данные в [Задании 16](#task16):

In [79]:
display(pd.pivot_table(data, index='total_income_category', values='debt', aggfunc=['sum', 'count', 'mean']))

sum  count      mean
                       debt   debt      debt
total_income_category                       
A                         2     25  0.080000
B                       354   5014  0.070602
C                      1353  15938  0.084891
D                        21    349  0.060172
E                         2     22  0.090909

**Вывод:** очевидно, что у людей с ежемесячным доходом меньше 30 тыс.₽ наблюдается и самый высокий процент задержки в выплатах (9,09%), на удивление, плохо отдают кредиты и самые обеспеченные граждане с доходом более 1 млн.₽ (8%). Самая же дисциплинированная категория имеет доход от 30 до 50 тыс.₽ у них процент просрочки 6,02%.

Самые массовые объёмы полученных кредитов наблюдаются в категорях "С" (объёмы доходов от 50'001₽ до 200'000₽) и "В" (объёмы доходов от 200'001₽ до 1'000'000₽). Вывод такой, что при увеличении доходов уменьшается количество задержек по выплатам. Но для полной уверенности, можно попробовать разбить эти категории на более узкие диапазоны доходов и посмотреть тендецию на большем количестве категорий (если пожелает заказчик или ревьювер в данном случае).

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

Для отображения данных по целям кредита и возврата в срок, воспользуемся столбцом "purpose_category", созданным нами в [Задании 18](#task18) и выведем сводную таблицу, сформированную по этим категориям:

In [80]:
display(pd.pivot_table(data, index='purpose_category', values='debt', aggfunc=['sum', 'count', 'mean']))

sum  count      mean
                         debt   debt      debt
purpose_category                              
операции с автомобилем    400   4281  0.093436
операции с недвижимостью  780  10754  0.072531
получение образования     369   3989  0.092504
проведение свадьбы        183   2324  0.078744

Это был способ через_pivot table, то же самое но через groupby():

In [81]:
display(data.groupby('purpose_category')['debt'].agg(['sum', 'count', 'mean']))

sum  count      mean
purpose_category                              
операции с автомобилем    400   4281  0.093436
операции с недвижимостью  780  10754  0.072531
получение образования     369   3989  0.092504
проведение свадьбы        183   2324  0.078744

Третий способ состоит в написании универсальной функции, в которую в качестве аргумента передаётся столбец, по которому будет производится группировка, а сама функция выполнит группировку, рассчитает и выведет на экран интересующие нас столбцы, саму функцию можно написать как на основе метода groupby() так и pivot_table:

In [97]:
def analyze_table(category):
    data_temp = data.pivot_table(index=category, values='debt', aggfunc=['count', 'sum', 'mean'])
    data_temp.columns = ['Количество кредитов', 'Количество должников', 'Доля должников']
    data_temp['Процент просрочек'] = round((data_temp['Доля должников'] * 100), 2)
    display (data_temp)

Далее вызываем эту функцию с интересующими нас столбцами в качестве аргумента и получаем все необходимые таблицы:

In [98]:
analyze_table('children')
analyze_table('family_status')
analyze_table('total_income_category')
analyze_table('purpose_category')

Количество кредитов  Количество должников  Доля должников  \
children                                                              
0                       14107                  1063        0.075353   
1                        4809                   444        0.092327   
2                        2052                   194        0.094542   
3                         330                    27        0.081818   
4                          41                     4        0.097561   
5                           9                     0        0.000000   

          Процент просрочек  
children                     
0                      7.54  
1                      9.23  
2                      9.45  
3                      8.18  
4                      9.76  
5                      0.00

Количество кредитов  Количество должников  \
family_status                                                      
Не женат / не замужем                 2796                   273   
в разводе                             1189                    84   
вдовец / вдова                         951                    63   
гражданский брак                      4146                   385   
женат / замужем                      12266                   927   

                       Доля должников  Процент просрочек  
family_status                                             
Не женат / не замужем        0.097639               9.76  
в разводе                    0.070648               7.06  
вдовец / вдова               0.066246               6.62  
гражданский брак             0.092861               9.29  
женат / замужем              0.075575               7.56

Количество кредитов  Количество должников  \
total_income_category                                              
A                                       25                     2   
B                                     5014                   354   
C                                    15938                  1353   
D                                      349                    21   
E                                       22                     2   

                       Доля должников  Процент просрочек  
total_income_category                                     
A                            0.080000               8.00  
B                            0.070602               7.06  
C                            0.084891               8.49  
D                            0.060172               6.02  
E                            0.090909               9.09

Количество кредитов  Количество должников  \
purpose_category                                                      
операции с автомобилем                   4281                   400   
операции с недвижимостью                10754                   780   
получение образования                    3989                   369   
проведение свадьбы                       2324                   183   

                          Доля должников  Процент просрочек  
purpose_category                                             
операции с автомобилем          0.093436               9.34  
операции с недвижимостью        0.072531               7.25  
получение образования           0.092504               9.25  
проведение свадьбы              0.078744               7.87

**Вывод:** самые дисциплинированные граждане в категории "операции с недвижимостью", видимо, не хотят рисковать своими квартирами и домами, процент задержек по выплатам 7,25%, те кто брал кредит на операции с автомобилями чаще всех задерживают выплаты (9,34%), так же большой процент задержек в категории "получение образования" (9,25%).

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:* причинами появления пропусков может быть человеский фактор, например кто-то забыл внести данные или случайно удалил и не заметил, а также это может быть результатом технического сбоя - ошибки в программном обеспечении при объединении каких-то таблиц или данных, сбои в результате отключения электричества или поломки оборудования.

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:* не всегда правильно заполнять пропуски медианным значением, иногда можно применить среднее арифметическое, бывают и другие методы, но когда некоторые значения сильно выделяются среди большинства, то для получения более достоверных сведений применяют медианное значение, например, среднюю температуру по больнице нельзя считать средним арифметическим, так как у большинства больных будет нормальная температура 36,6 и лишь у тех немногих, у кого воспалительные процессы температура будет повышенной.

## Общий вывод.

Безусловно все рассмотренные категории имеют влияние на возврат кредита в срок. Если суммировать данные по этим категориям, то самая рискованная категория граждан для выдачи кредита будет "многодетная незамужняя с четырьмя детьми, доходом менее 30 тыс.₽, которая хочет купить автомобиль". А если серьёзно, то по полученным данным можно сформировать весовые коэффициенты, так как у нас теперь есть численные данные по конкретным категориям и автоматически считать риски при решении о выдаче кредита тому или иному гражданину.

Главной целью данной работы было исследование надёжности заёмщиков. Для этого были предоставлены анкетные данные о количестве детей в семье, общем трудовом стаже, возрасте, уровне образования, семейном положении, поле клиента, типе занятости, имелась ли задолженность по кредитам, ежемесячный доход и цели получения кредита.

В процессе изучения данных выявились пропуски в данных об общем трудовом стаже и ежемесячном доходе. Пропущенные данные о доходах для людей с типом занятости "сотрудник" были заполнены медианным значением доходов для этой категории граждан. С данными о трудовом стаже исследования не проводились, тем не менее, отрицательные значения заменены на положительные, пропущенные на соответствующие медианные значения для каждой категории занятости. В данных по количеству детей были удалены строки с аномальными данными. Следующим важным шагом были удалены найденные строки-дубликаты. Все клиенты были разделены на 5 категорий по уровню доходов для того, чтобы было можно качественно оценить задолженности в каждой категории. Для этого же все цели получения кредита были разделены на четыре простые и понятные категории.
    
Далее мы провели исследования на зависимость количества детей, семейного положения, уровня дохода и цели получения кредита на возврат кредита в срок. В результате выяснили, что все эти данные имеют влияние. Чем больше детей, тем выше риски просроченных платежей, состоящие или бывшие в браке заемщики более дисциплинированы, при увеличении доходов уменьшается количество задержек по выплатам, а самыми проблемными выглядят автокредиты и что удивительно на образование.
    
Так как большинство пропусков, ошибок и аномальных данных появлется в результате человечекого фактора можно было бы улучшить программное обеспечение для ввода анкетных данных. Например, добавить проверку на ввод отрицательных значений при заполении данных о детях и количестве дней трудового стажа. Для данных о доходах можно сделать автоматическое присвоение категорий (по количеству и значениям удобных именно для банка), также можно категорезировать цели кредита, чтобы оператор выбирал их из уже готового списка. Все эти меры значительно уменьшат количество ошибок в данных, упростят и ускорят их анализ и обработку, а также сделают результаты более точными.
    
По полученным данным можно сформировать весовые коэффициенты и автоматически считать риски при решении о выдаче кредита тому или иному гражданину.